tools/fp20compiler/main.cpp

```diff
@@ -37,25 +37,69 @@ char * myin = 0;
 void translate(const char* s) {
     char* ns = strdup(s);
 
-    // printf("-- %s\b", s);
-
-    char* ts_start = strstr(ns, "!!TS1.0");
-    if (ts_start) {
-        char* ts_end = strstr(ts_start, "// End of program");
-        assert(ts_end);
-        *ts_end = 0;
-
-        ts10_init(ts_start);
-        ts10_parse();
-
-        *ts_end = '/';
-    }
-
-    char* rc_start = strstr(ns, "!!RC1.0");
-    if (rc_start) {
-        rc10_init(rc_start);
-        rc10_parse();
-    }
+    // Move to start of file
+    char* cursor = ns;
+
+    // Remember where the current shader starts
+    char* shader_start = NULL;
+
+    // Loop over each shader (between two shader start tokens)
+    do {
+
+        // Try to find a shader start token candidate to mark the end
+        char* shader_end = strstr(cursor, "!!");
```

Comment:
This is misleading. It doesn't search a `shader_end`, but `next_shader_start`.
It's used as `shader_start` or `shader_end` depending on the current state:
- If we don't have a start for the current shader, the next shader start marks the start
- If we already have a start for the current shader, the next shader start marks the end

Reply:
Fixed. I rewrote the algorithm differently.